libraries/AP_KDECAN/AP_KDECAN.cpp

```diff
@@ -53,603 +41,282 @@ const AP_Param::GroupInfo AP_KDECAN::var_info[] = {
+            {
+                const uint8_t idx = id.source_id - ESC_NODE_ID_FIRST;
+                const uint8_t num_poles = _telemetry.num_poles > 0 ? _telemetry.num_poles : DEFAULT_NUM_POLES;
+                update_rpm(idx, uint16_t(uint16_t(frame.data[4] << 8 | frame.data[5]) * 60UL * 2 / num_poles));
+
+                TelemetryData t {
+                    .temperature_cdeg = int16_t(frame.data[6] * 100),
+                    .voltage = float(uint16_t(frame.data[0] << 8 | frame.data[1])) * 0.01f,
+                    .current = float(uint16_t(frame.data[2] << 8 | frame.data[3])) * 0.01f,
+                };
+                update_telem_data(idx, t,
+                    AP_ESC_Telem_Backend::TelemetryType::CURRENT |
+                    AP_ESC_Telem_Backend::TelemetryType::VOLTAGE |
+                    AP_ESC_Telem_Backend::TelemetryType::TEMPERATURE);
+            }
+            break;
+#endif // HAL_WITH_ESC_TELEM
 
-    AP_HAL::CANFrame frame { (id.value | AP_HAL::CANFrame::FlagEFF), nullptr, 0 };
+        case UPDATE_NODE_ID_OBJ_ADDR:
+            // reply from setting new node ID
+            _init.detected_bitmask |= 1U << (id.source_id - ESC_NODE_ID_FIRST);
+            GCS_SEND_TEXT(MAV_SEVERITY_INFO, "KDECAN: Found ESC id %u", id.source_id);
+            break;
+    }
+}
 
-    if(!_can_iface->send(frame, AP_HAL::micros() + 1000000, 0)) {
-        debug_can(AP_CANManager::LOG_DEBUG, "couldn't send discovery message");
+void AP_KDECAN_Driver::update(const uint8_t num_poles)
+{
+    if (!hal.util->get_soft_armed() || _init.detected_bitmask == 0) {
         return;
     }
 
-    debug_can(AP_CANManager::LOG_DEBUG, "discovery message sent");
-
-    uint32_t start = AP_HAL::millis();
-
-    // wait 1 second for answers
-    while (AP_HAL::millis() - start < 1000) {
-        AP_HAL::CANFrame esc_id_frame {};
-        uint64_t rx_time;
-        AP_HAL::CANIface::CanIOFlags flags = 0;
-
-        int16_t n = _can_iface->receive(esc_id_frame, rx_time, flags);
-
-        if (n != 1) {
-            continue;
-        }
-
-        if (!esc_id_frame.isExtended()) {
-            continue;
-        }
+#if HAL_WITH_ESC_TELEM
+    _telemetry.num_poles = num_poles;
+#endif
+    
+    uint16_t pwm[KDECAN_MAX_NUM_ESCS] {};
 
-        if (esc_id_frame.dlc != 5) {
+    uint8_t index = 0;
+    for (uint16_t i = 0; i < ARRAY_SIZE(pwm); i++) {
+        if ((_init.detected_bitmask & (1UL<<index)) == 0) {
             continue;
         }
-
-        id.value = esc_id_frame.id & AP_HAL::CANFrame::MaskExtID;
-
-        if (id.source_id == BROADCAST_NODE_ID ||
-            id.source_id >= (KDECAN_MAX_NUM_ESCS + ESC_NODE_ID_FIRST) ||
-            id.destination_id != AUTOPILOT_NODE_ID ||
-            id.object_address != ESC_INFO_OBJ_ADDR) {
-            continue;
+        if (SRV_Channels::channel_function(i) > SRV_Channel::Aux_servo_function_t::k_none) {
+            pwm[index++] = SRV_Channels::srv_channel(i)->get_output_pwm();
```

Comment:
Potential nullptr dereference, depending on whatever srv_channel decides to do.  You'll definitely get one if someone reduces num-serv-channels.

Reply:
Fixed